### PR TITLE
Parser tests

### PR DESCRIPTION
**Pull request by Razican**
Hi!

This is my first contribution to the project, great work! I'm working on doing test for the parser (#40), and I've just created the first changes. Note that I ran `cargo fmt`, so there are minor format changes in other places.

I wanted to create the PR asap, so that you could review it and let me know if I'm following the proper procedure. Real changes are in the `src/lib/syntax/parser.rs` file, where I added a `tests` module. For now, it just contains tests for strings, arrays and declarations.

I found some errors in the parser, though:
 - Trailing commas in array assignments make the parser fail ([spec link][array_assignment]).
 - `const` declarations work with empty expressions when they shouldnt ([spec_link][empty_const]).

For this last error, I would suggest to change the definition of `syntax::ast::expr::ExprDef::ConstDeclExpr` to directly not allow an `Option` in the expression.

Let me know if you like the design, and I'll implement the tests for the rest of the expressions!

[array_assignment]: https://tc39.es/ecma262/#prod-ArrayAssignmentPattern
[empty_const]: https://tc39.es/ecma262/#sec-let-and-const-declarations-static-semantics-early-errors

### CODE REVIEW

**Comment by jasonwilliams** [comment]
That's aweome @Razican i will take a look, thanks

File: src/lib/syntax/parser.rs [hunk 1]
```diff
@@ -757,3 +757,180 @@ impl Parser {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::syntax::{
+        ast::expr::{Expr, ExprDef},
+        lexer::Lexer,
+    };
+
+    fn check_parser(js: &str, expr: &[Expr]) {
```
Comment (jasonwilliams) [inline src/lib/syntax/parser.rs]: I like this idea, makes the further tests much cleaner and easier to understand

**Comment by jasonwilliams** [comment]
Looks like some contributors didn't have cargo fmt on when they PR'd, i wonder if there's a way to enforce this

**Comment by Razican** [comment]
> Looks like some contributors didn't have cargo fmt on when they PR'd, i wonder if there's a way to enforce this

Actually there is!
I use rustfmt in CI for my other repos, such as [SUPER](https://github.com/SUPERAndroidAnalyzer/super)

I just add the rustup component in Travis CI and then check the format: [travis-helper.sh#L23](https://github.com/SUPERAndroidAnalyzer/super/blob/master/travis-helper.sh#L23)

I fixed the links to the docs in the PR, by the way, I forgot to add them.

**Comment by jasonwilliams** [comment]
That's good to know!

**Comment by jasonwilliams** [comment]
Yes i like the design, carry on, it looks good so far

**Comment by jasonwilliams** [comment]
> I found some errors in the parser, though:
> - Trailing commas in array assignments make the parser fail ([spec link][array_assignment]).
> - `const` declarations work with empty expressions when they shouldnt ([spec_link][empty_const]).
>
>For this last error, I would suggest to change the definition of `syntax::ast::expr::ExprDef::ConstDeclExpr` to directly not allow an `Option` in the expression.

Issues can be raised for these

**Comment by jasonwilliams** [comment]
It might be an idea to merge what you have so far to get Travis working properly again

**Comment by Razican** [comment]
> It might be an idea to merge what you have so far to get Travis working properly again.
> Let me know if you're happy to do that

Sure, give me until the end of the day (CEST time) to push the last tests I did parsing numeric binary operations and we can merge, and I can keep working on it later.

By the way, is there an specification on how should the parser parse complex mathematical operations? For example:

`a + d*(b-3) + 1`

I see that it will generate something like this in the parser:

`a + (1 + (d * (b-3)))`

Is there any reason why it changes the order of the last addition?

**Comment by jasonwilliams** [comment]
I think the order here might just be wrong: https://github.com/jasonwilliams/boa/blob/master/src/lib/syntax/parser.rs#L725-L729

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence

Can't see anything in the spec

**Comment by Razican** [comment]
I pushed another commit with more parser tests. I set up a small function to create binary operations and check numeric binary operators.

**Comment by jasonwilliams** [comment]
latest commit is failing on some code formatting

**Comment by Razican** [comment]
Fixed, it seems that when I rebased there was some formatting issue.

File: src/lib/syntax/parser.rs [hunk 2]
```diff
@@ -757,3 +757,354 @@ impl Parser {
...
+        let mut lexer = Lexer::new(js);
+        lexer.lex().unwrap();
+
+        dbg!(Parser::new(lexer.tokens).parse_all().unwrap());
```
Comment (jasonwilliams) [inline src/lib/syntax/parser.rs]: Is there a reason this is wrapped in `dbg!` ?

Comment (Razican) [reply src/lib/syntax/parser.rs]: Sorry, it was left there from my debugging. It's removed now, and I added bitwise operators.

**Comment by Razican** [comment]
Hi, I think this is ready for merge now.

**Pull request merged**